src/python/pants/rules/core/cloc.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pathlib import Path
+from typing import Set
+
+from pants.base.build_environment import get_buildroot
+from pants.base.specs import Specs
+from pants.engine.console import Console
+from pants.engine.fs import (Digest, DirectoriesToMerge, FilesContent, InputFileContent, Snapshot,
+                             UrlToFetch)
+from pants.engine.goal import Goal, LineOriented
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.legacy.graph import HydratedTargets, TransitiveHydratedTargets
+from pants.engine.rules import console_rule, rule
+from pants.engine.selectors import Get
+from pants.util.objects import datatype
+
+
+class DownloadedClocScript(datatype([('script_path', str), ('digest', Digest)])):
+  """Cloc script as downloaded from the pantsbuild binaries repo."""
+
+
+@rule(DownloadedClocScript, [])
+def download_cloc_script():
+  url = "https://binaries.pantsbuild.org/bin/cloc/1.80/cloc"
+  sha_256 = "2b23012b1c3c53bd6b9dd43cd6aa75715eed4feb2cb6db56ac3fbbd2dffeac9d"
+  digest = Digest(sha_256, 546279)
+  snapshot = yield Get(Snapshot, UrlToFetch(url, digest))
+  yield DownloadedClocScript(script_path=snapshot.files[0], digest=snapshot.directory_digest)
+
+
+class CountLinesOfCode(LineOriented, Goal):
+  name = 'cloc'
+
+  @classmethod
+  def register_options(cls, register) -> None:
+    super().register_options(register)
+    register('--transitive', type=bool, fingerprint=True, default=True,
+             help='Operate on the transitive dependencies of the specified targets.  '
+                  'Unset to operate only on the specified targets.')
+    register('--ignored', type=bool, fingerprint=True,
+             help='Show information about files ignored by cloc.')
+
+
+@console_rule(CountLinesOfCode, [Console, CountLinesOfCode.Options, DownloadedClocScript, Specs])
+def run_cloc(console, options, cloc_script, specs):
+  """Runs the cloc perl script in an isolated process"""
+
+  transitive = options.values.transitive
+  ignored = options.values.ignored
+
+  if (transitive):
+    targets = yield Get(TransitiveHydratedTargets, Specs, specs)
+    all_target_adaptors = {t.adaptor for t in targets.closure}
+  else:
+    targets = yield Get(HydratedTargets, Specs, specs)
+    all_target_adaptors = {t.adaptor for t in targets}
+
+  build_root = get_buildroot()
+  source_paths: Set[str] = set()
+  for t in all_target_adaptors:
+    sources = getattr(t, 'sources', None)
+    if sources is not None:
+      for f in sources.snapshot.files:
+        source_paths.add(str(Path(build_root, f)))
```

Comment:
These are absolute, which means that the rule will escape its sandbox. Instead, they should be relative, so that the contents of the snapshot actually get used.